File: Tree/medium/98.cpp
// Runtime: 16 ms, faster than 55.22% of C++ online submissions for Validate Binary Search Tree.
// Memory Usage: 18.4 MB, less than 97.92% of C++ online submissions for Validate Binary Search Tree.

// https://leetcode.com/problems/validate-binary-search-tree/discuss/32109/My-simple-Java-solution-in-3-lines

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return recursive(root, LONG_MIN, LONG_MAX);
    }
    
    bool recursive(TreeNode* root, long minVal, long maxVal) {
        if (root == NULL) return true;
        if (minVal >= root->val || maxVal <= root->val) return false;
        return recursive(root->left, minVal, root->val) && recursive(root->right, root->val, maxVal);
    }
};

File: Tree/medium/98.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        def helper(curr: TreeNode, lowerBound: int, upperBound: int):
            if curr == None:
                return True

            if (lowerBound != None and lowerBound >= curr.val) or (upperBound != None and curr.val >= upperBound):
                return False

            return helper(curr.left, lowerBound, curr.val) and helper(curr.right, curr.val, upperBound)

        return helper(root, None, None)


